### ISSUE TITLE
Uncaught error in elk master (possible regression from elkjs@0.4.1)

### ISSUE THREAD

**Issue by jbeard4** [kieler/elkjs#48]
I believe there is a regression in elk master. Please consider the following reduced example:

Expected output (generated in elkjs@0.4.1):

<img width="826" alt="screen shot 2018-08-16 at 11 15 41 am" src="https://user-images.githubusercontent.com/586962/44217642-ef099c80-a145-11e8-9a7c-aea9c28fbd60.png">

Actual output:

An uncaught error is thrown:
`localhost/:1 Uncaught (in promise) java.lang.IndexOutOfBoundsException: Index: 2, Size: 2`

Here is kgraph before layout:

```json
{
    "id": "$generated-scxml-0",
    "labels": [
        {
            "text": ""
        }
    ],
    "edges": [
        {
            "id": "A:transition:0",
            "source": "A",
            "target": "a1",
            "labels": [],
            "sourcePort": "A_enter:0",
            "targetPort": "a1_enter:0",
            "$hyperlink": "A_exit:1_A_enter:0"
        },
        {
            "id": "$generated_A_initial_0:transition:0",
            "source": "$generated_A_initial_0",
            "target": "a1",
            "labels": []
        },
        {
            "id": "$generated_$generated-scxml-0_initial_0:transition:0",
            "source": "$generated_$generated-scxml-0_initial_0",
            "target": "A",
            "labels": []
        },
        {
            "id": "A_exit:1_A_enter:0",
            "source": "A",
            "target": "A",
            "sourcePort": "A_exit:1",
            "targetPort": "A_enter:0",
            "labels": [
                {
                    "text": "e",
                    "width": 1.765625,
                    "height": 4.53125
                }
            ],
            "$type": "hyperlink"
        }
    ],
    "properties": {
        "borderSpacing": 6
    },
    "$type": "scxml",
    "children": [
        {
            "id": "A",
            "labels": [
                {
                    "text": "A"
                }
            ],
            "edges": [],
            "properties": {
                "borderSpacing": 6
            },
            "$type": "state",
            "children": [
                {
                    "id": "a1",
                    "labels": [
                        {
                            "text": "a1"
                        }
                    ],
                    "edges": [],
                    "properties": {
                        "borderSpacing": 6
                    },
                    "$type": "state",
                    "ports": [
                        {
                            "id": "a1_enter:0"
                        },
                        {
                            "id": "a1_exit:1"
                        }
                    ],
                    "x": 0,
                    "y": 0,
                    "width": 8.765625,
                    "height": 9.53125
                },
                {
                    "id": "a2",
                    "labels": [
                        {
                            "text": "a2"
                        }
                    ],
                    "edges": [],
                    "properties": {
                        "borderSpacing": 6
                    },
                    "$type": "state",
                    "x": 0,
                    "y": 0,
                    "width": 8.765625,
                    "height": 9.53125
                },
                {
                    "id": "$generated_A_initial_0",
                    "labels": [
                        {
                            "text": "◉ $generated_A_initial_0"
                        }
                    ],
                    "edges": [],
                    "properties": {
                        "borderSpacing": 6
                    },
                    "$type": "initial",
                    "x": 0,
                    "y": 0,
                    "width": 4,
                    "height": 4
                }
            ],
            "ports": [
                {
                    "id": "A_enter:0"
                },
                {
                    "id": "A_exit:1"
                }
            ],
            "x": 0,
            "y": 0,
            "width": 7.875,
            "height": 9.53125
        },
        {
            "id": "$generated_$generated-scxml-0_initial_0",
            "labels": [
                {
                    "text": "◉ $generated_$generated-scxml-0_initial_0"
                }
            ],
            "edges": [],
            "properties": {
                "borderSpacing": 6
            },
            "$type": "initial",
            "x": 0,
            "y": 0,
            "width": 4,
            "height": 4
        }
    ],
    "width": 5,
    "height": 5,
    "layoutOptions": {
        "org.eclipse.elk.hierarchyHandling": "INCLUDE_CHILDREN",
        "org.eclipse.elk.algorithm": "org.eclipse.elk.layered",
        "org.eclipse.elk.padding": 0,
        "org.eclipse.elk.nodeLabels.padding": 0
    }
}
```

This might be related to the new self-loop label placement code.

Thank you for your help with this.

**Comment by jbeard4** [comment]
I don't know if it makes sense to report issues in master, as this code is not yet part of a public release, and so is more likely to have bugs. 

I also think it would probably be easier to try to reproduce and fix this error in elk, rather than elkjs, as the error provided by elkjs pretty opaque. I have hopped on the elk-dev mailing list and am planning to set up elk for development, then see if I can reproduce and resolve in elk. Thanks.

**Comment by le-cds** [comment]
> I don't know if it makes sense to report issues in master
It does, please keep the reports coming.

**Comment by jbeard4** [comment]
Thanks, will do.

**Comment by jbeard4** [comment]
I wanted to let you know, I tested the latest elk commit e9ae57e2d92327c8ae42aab39f1aa3fea7ecc913 because it looked like it might fix the issue, but it is still recurring. 

Thanks for your help with this.

**Comment by le-cds** [comment]
It's the self loop label placement code that seems to be causing the issue. I'll create a ticket at the main ELK project.

**Comment by le-cds** [comment]
See eclipse/elk#368.

**Comment by le-cds** [comment]
@uruuru is it okay in JSON to specify edges that way? My feeling is that they have the wrong containment. When I rebuilt the example in an `elkt` file, I did have to change the containment in order to get the example to even run.

**Comment by uruuru** [comment]
In theory the containment should be respected as by our conventions. IIRC the edge containment of `elkt` files is adapted automatically in the underlying elk graph - not so in the json format.

**Comment by le-cds** [comment]
> IIRC the edge containment of elkt files is adapted automatically in the underlying elk graph - not so in the json format.

Could you explain that a bit more? I'm not sure I got your point.

**Comment by uruuru** [comment]
I believe there's code in the elkt parser code that makes sure an edge is placed at its right containment (including showing a warning in the textual editor). This is not the case for elkjs - you have to make sure yourself that the edge is specified at the correct containment.

**Comment by uruuru** [comment]
The edge containments are, in fact, off. This keeps occurring ... 

I cleaned the example a bit and it seems to be fine wrt. to the problem discussed in this issue. There's another issue that keeps it from looking completely fine that will be fixed for 0.7.0. 

```
{
  "id": "root",
  "children": [
    {
      "id": "A",
      "labels": [ { "text": "A" } ],
      "children": [
        {
          "id": "a1",
          "labels": [ { "text": "a1" } ],
          "ports": [
            { "id": "a1_enter" },
            { "id": "a1_exit" }
          ],
          "width": 20,
          "height": 20
        },
        {
          "id": "a2",
          "labels": [ { "text": "a2" } ],
          "width": 20,
          "height": 20
        },
        {
          "id": "a3",
          "labels": [ { "text": "a3" } ],
          "width": 20,
          "height": 20
        }
      ],
      edges: [
        { "id": "e4", "source": "a3", "target": "a1" },
        { "id": "e3", "source": "A", "target": "a1", "sourcePort": "A_enter", "targetPort": "a1_enter" }
      ],
      "ports": [
        { "id": "A_enter" },
        { "id": "A_exit" }
      ],
      "width": 7.875,
      "height": 9.23253125
    },
    {
      "id": "B",
      "labels": [ { "text": "B" } ],
      "width": 20,
      "height": 20
    }
  ],
  "edges": [
    { "id": "e2", "source": "B", "target": "A" },
    // self loop
    { "id": "e1", 
      "source": "A", "target": "A", 
      "sourcePort": "A_exit", "targetPort": "A_enter", 
      "labels": [ { "text": "e" } ] 
    }
  ],
  "layoutOptions": {
    "org.eclipse.elk.hierarchyHandling": "INCLUDE_CHILDREN",
    "crossingMinimization.greedySwitch.type": "OFF",
    "org.eclipse.elk.algorithm": "org.eclipse.elk.layered",
  }
}
```
